Add test coverage to `actions/create_realm.py` and `actions/message_edit.py` files.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is to remove unused code.
- Second commit adds coverage to `actions/create_realm.py`.
- Third commit adds coverage to `actions/message_edit.py`.
- Fourth commit is to add "nocoverage" comment from unreachable code.

 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on zerver/tests/test_message_edit.py:
```diff
@@ -2288,6 +2299,51 @@ def test_moving_all_topic_messages(
+
+        # Check message count in new topic and/or stream.
+        messages = get_topic_messages(user_profile, new_stream, new_topic)
+        self.assert_length(messages, 5)
```
I am not sure whether it is worth to add tests for a topic-only and stream-only edit for this case when there is no time limit set.

timabbott replied:
Yeah it's probably unnecessary.

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -918,7 +918,7 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
                 # This corresponds to the case when messages are moved
                 # to a stream-topic pair that didn't exist.
-                if new_visibility_policy == UserTopic.VisibilityPolicy.INHERIT:
+                if new_visibility_policy == UserTopic.VisibilityPolicy.INHERIT:  # nocoverage
```
I have tried to explain this in the commit message, but would mention here at weill with code.

There will be no UserTopic rows for original topic with visibility set to `INHERIT` since we delete them.

And since the target topic has no messages, there will be no UserTopic rows for it. So, the code below also does not add any entry to `orig_topic_user_profile_to_visibility_policy`.
```python
user_profiles_having_visibility_policy: Set[UserProfile] = set(
    itertools.chain(
        orig_topic_user_profile_to_visibility_policy.keys(),
         target_topic_user_profile_to_visibility_policy.keys(),
     )
)

user_profiles_for_visibility_policy_pair: Dict[
    Tuple[int, int], List[UserProfile]
] = defaultdict(list)
for user_profile_with_policy in user_profiles_having_visibility_policy:
    if user_profile_with_policy not in target_topic_user_profile_to_visibility_policy:
        target_topic_user_profile_to_visibility_policy[
            user_profile_with_policy
        ] = UserTopic.VisibilityPolicy.INHERIT
    elif user_profile not in orig_topic_user_profile_to_visibility_policy:
        orig_topic_user_profile_to_visibility_policy[
            user_profile_with_policy
        ] = UserTopic.VisibilityPolicy.INHERIT

        orig_topic_visibility_policy = orig_topic_user_profile_to_visibility_policy[
            user_profile_with_policy
        ]
        target_topic_visibility_policy = target_topic_user_profile_to_visibility_policy[
             user_profile_with_policy
        ]
        user_profiles_for_visibility_policy_pair[
             (orig_topic_visibility_policy, target_topic_visibility_policy)
        ].append(user_profile_with_policy)
``` 

So, there is actually no entry in `user_profiles_for_visibility_policy_pair` with `orig_topic_visibility_policy` set to `INHERIT`.

sahil839 replied:
I am not sure whether we want to remove the code itself or just keep it for safety purposes to avoid unnecessary operations.

sahil839 replied:
@prakhar1144 FYI. Can you confirm if I am missing something here?

sahil839 replied:
> adds the entry in user_profiles_for_visibility_policy_pair with orig_topic_visibility_policy set to INHERIT.
Ok, I got this. But the below code will run only if the user_profile exists in `user_profiles_having_visibility_policy` and does not exist in `orig_topic_user_profile_to_visibility_policy`. So, it needs to exist in `target_topic_user_profile_to_visibility_policy`, but it cannot since there are no messages in the target topic (essentially meaning no topic exists) and hence there should not be any `UserTopic` rows for it, correct?

```python
elif user_profile not in orig_topic_user_profile_to_visibility_policy:
       orig_topic_user_profile_to_visibility_policy[
           user_profile_with_policy
       ] = UserTopic.VisibilityPolicy.INHERIT
```

prakhar1144 replied:
In that case, we are dealing with the set of users
target_topic: INHERIT (No UserTopic row)
original_topic: INHERIT (No UserTopic row)

so, user_profile neither exists in `orig_topic_user_profile_to_visibility_policy` nor `target_topic_user_profile_to_visibility_policy` hence not in `user_profiles_having_visibility_policy`.


And we don't need to handle the "move user topic records" in this case, simply move the message. For those users, target_topic visibility policy is already INHERIT (i.e no UserTopic row exists).

-----
Also, is CZO down? -- can you please confirm. It just keeps loading for me.

sahil839 replied:
Yeah so, we cannot reach this if-conditon then.

> Also, is CZO down? -- can you please confirm. It just keeps loading for me.

yes same for me.

prakhar1144 replied:
> Yeah so, we cannot reach this if-conditon then.

if
target_topic has no messages (hence, no UserTopic row)
orig_topic: MUTED

Now,
`orig_topic_user_profile_to_visibility_policy` has user_profile
`user_profiles_having_visibility_policy` has user_profile
```python
if user_profile_with_policy not in target_topic_user_profile_to_visibility_policy:
        target_topic_user_profile_to_visibility_policy[
            user_profile_with_policy
        ] = UserTopic.VisibilityPolicy.INHERIT
```

Now, the if block will be reached, right?

prakhar1144 replied:
Ah, right. -- that can be removed.

timabbott replied:
I think it's possible to have a `UserTopic` row for a topic with no messages. I think the case for this code path is a message is sent to a topic, and then you mute the topic, and then someone deletes the message, and then someone tries to move messages from a topic that you've unmuted in some other stream into the (empty of messages, but muted) topic name.

prakhar1144 replied:
Hm, I can [recall a discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/duplicate.20request.20to.20update.20UserTopic.20row/near/1523816) around this on CZO.

> So in short there should be no requirement that the topic already have messages in it when creating a UserTopic record.

sahil839 replied:
Hmm, I agree there can be such case. But in such cases as well, the current code seems wrong. Because we just skip the case when the `new_visibility_policy` is `INHERIT`, and I think we do so assuming there is no `UserTopic` row for a topic with no messages. 

I have updated the code and added the tests too. Will push to this PR soon.

sahil839 commented:
Updated the PR.
- Added commit to fix the case of having UserTopic rows for a topic without messages in case of deleted messages as discussed [above in the comments](https://github.com/zulip/zulip/pull/25508#discussion_r1190364625). Also fixed a typo in that commit.
- Added tests to test the case of target topic without messages.

timabbott commented on zerver/tests/test_message_edit.py:
```diff
@@ -1893,6 +1893,149 @@ def assert_has_visibility_policy(
+        self.subscribe(cordelia, stream_name)
+        self.subscribe(aaron, stream_name)
+
+        def assert_has_visibility_policy(
```
We might want to consider making this a function not defined in 3 different individual tests

timabbott commented:
This is great, merged, thanks @sahil839! Definitely one of those nice cases where trying to fix missing test coverage helped us find a live bug :)
